File: android/lib/services/dynamodb_service.dart
import 'package:aws_dynamodb_api/dynamodb-2012-08-10.dart';

import '../models/obsyncian_config.dart';
import 'aws_signing_client.dart';

/// Represents a sync timestamp item in the Obsyncian DynamoDB table.
class SyncItem {
  final String userId;
  final String timestamp;

  const SyncItem({required this.userId, required this.timestamp});
}

/// Service for tracking sync timestamps in DynamoDB.
///
/// Mirrors the Go app's DynamoDB operations:
///   - Scan for latest sync across all users
///   - Get/create/update individual user entries
///
/// The DynamoDB table ("Obsyncian") has a primary key of "UserId" (String)
/// and a "Timestamp" attribute (String, format "YYYYMMDDHHmmss").
class DynamoDBService {
  static const tableName = 'Obsyncian';

  late DynamoDB _client;
  AwsSigningClient? _signingClient;

  DynamoDBService();

  void configure(ObsyncianConfig config) {
    // Close previous signing client if any
    _signingClient?.close();

    // Use the official AWS SigV4 signer via a custom HTTP client
    _signingClient = AwsSigningClient(
      accessKey: config.credentials.key,
      secretKey: config.credentials.secret,
      region: config.awsRegion,
      service: 'dynamodb',
    );

    _client = DynamoDB(
      region: config.awsRegion,
      credentials: AwsClientCredentials(
        accessKey: config.credentials.key,
        secretKey: config.credentials.secret,
      ),
      client: _signingClient!,
    );
  }

  /// Scan the table to find the latest sync timestamp across all users.
  /// Returns null if the table is empty.
  Future<SyncItem?> getLatestSync() async {
    final allItems = <SyncItem>[];
    Map<String, AttributeValue>? lastEvaluatedKey;

    do {
      final result = await _client.scan(
        tableName: tableName,
        projectionExpression: 'UserId, #ts',
        expressionAttributeNames: {'#ts': 'Timestamp'},
        exclusiveStartKey: lastEvaluatedKey,
      );

      if (result.items != null) {
        for (final item in result.items!) {
          final userId = item['UserId']?.s;
          final timestamp = item['Timestamp']?.s;
          if (userId != null && timestamp != null) {
            allItems.add(SyncItem(userId: userId, timestamp: timestamp));
          }
        }
      }

      lastEvaluatedKey = result.lastEvaluatedKey;
    } while (lastEvaluatedKey != null && lastEvaluatedKey.isNotEmpty);

    if (allItems.isEmpty) return null;

    // Sort descending by timestamp, take the latest
    allItems.sort((a, b) => b.timestamp.compareTo(a.timestamp));
    return allItems.first;
  }

  /// Get a specific user's item from the table. Returns null if not found.
  Future<SyncItem?> getUser(String userId) async {
    final result = await _client.getItem(
      tableName: tableName,
      key: {'UserId': AttributeValue(s: userId)},
    );

    if (result.item == null) return null;

    final ts = result.item!['Timestamp']?.s;
    if (ts == null) return null;

    return SyncItem(userId: userId, timestamp: ts);
  }

  /// Create a new user entry with the current timestamp.
  Future<void> createUser(String userId) async {
    final timestamp = _nowTimestamp();
    await _client.putItem(
      tableName: tableName,
      item: {
        'UserId': AttributeValue(s: userId),
        'Timestamp': AttributeValue(s: timestamp),
      },
    );
  }

  /// Update the user's timestamp to now.
  Future<void> updateTimestamp(String userId) async {
    final timestamp = _nowTimestamp();
    await _client.updateItem(
      tableName: tableName,
      key: {'UserId': AttributeValue(s: userId)},
      updateExpression: 'SET #ts = :ts',
      expressionAttributeNames: {'#ts': 'Timestamp'},
      expressionAttributeValues: {':ts': AttributeValue(s: timestamp)},
      returnValues: ReturnValue.updatedNew,
    );
  }

  /// Generate a timestamp string in the format used by the Go app: "YYYYMMDDHHmmss".
  /// Uses local time (not UTC) to match the Go app's `time.Now().Format(...)`.
  String _nowTimestamp() {
    final now = DateTime.now();
    return '${now.year.toString().padLeft(4, '0')}'
        '${now.month.toString().padLeft(2, '0')}'
        '${now.day.toString().padLeft(2, '0')}'
        '${now.hour.toString().padLeft(2, '0')}'
        '${now.minute.toString().padLeft(2, '0')}'
        '${now.second.toString().padLeft(2, '0')}';
  }
}

File: android-rn/src/services/dynamoDBService.ts
import {
  DynamoDBClient,
  ScanCommand,
  GetItemCommand,
  PutItemCommand,
  UpdateItemCommand,
} from '@aws-sdk/client-dynamodb';
import { ObsyncianConfig } from '../models/config';

const TABLE_NAME = 'Obsyncian';

export interface UserSync {
  userId: string;
  timestamp: string; // YYYYMMDDHHmmss (UTC)
}

class DynamoDBService {
  private client: DynamoDBClient | null = null;
  private lastLog: (msg: string) => void = () => {};

  async init(config: ObsyncianConfig, onLog: (msg: string) => void): Promise<void> {
    this.lastLog = onLog;
    this.client = new DynamoDBClient({
      region: config.region,
      credentials: {
        accessKeyId: config.credentials.key,
        secretAccessKey: config.credentials.secret,
      },
    });
  }

  async getLatestSync(): Promise<UserSync | null> {
    if (!this.client) throw new Error('DynamoDBService not initialized');

    try {
      this.lastLog('[DynamoDB] Scanning for latest timestamp');
      let latest: UserSync | null = null;
      let exclusiveStartKey: any;

      while (true) {
        const response = await this.client.send(
          new ScanCommand({
            TableName: TABLE_NAME,
            ExclusiveStartKey: exclusiveStartKey,
          }),
        );

        if (response.Items) {
          for (const item of response.Items) {
            const userId = item.UserId?.S;
            const timestamp = item.Timestamp?.S;

            if (userId && timestamp) {
              if (!latest || timestamp > latest.timestamp) {
                latest = { userId, timestamp };
              }
            }
          }
        }

        if (!response.LastEvaluatedKey) break;
        exclusiveStartKey = response.LastEvaluatedKey;
      }

      if (latest) {
        this.lastLog(`[DynamoDB] Latest sync: ${latest.userId} @ ${latest.timestamp}`);
      } else {
        this.lastLog('[DynamoDB] Table is empty');
      }

      return latest;
    } catch (error) {
      this.lastLog(`[DynamoDB] getLatestSync error: ${error}`);
      throw error;
    }
  }

  async getUser(userId: string): Promise<UserSync | null> {
    if (!this.client) throw new Error('DynamoDBService not initialized');

    try {
      const response = await this.client.send(
        new GetItemCommand({
          TableName: TABLE_NAME,
          Key: {
            UserId: { S: userId },
          },
        }),
      );

      if (response.Item?.Timestamp?.S) {
        return {
          userId,
          timestamp: response.Item.Timestamp.S,
        };
      }

      return null;
    } catch (error) {
      this.lastLog(`[DynamoDB] getUser error: ${error}`);
      throw error;
    }
  }

  async createUser(userId: string): Promise<void> {
    if (!this.client) throw new Error('DynamoDBService not initialized');

    const timestamp = this.nowTimestamp();
    this.lastLog(`[DynamoDB] Creating user ${userId} with timestamp ${timestamp}`);

    try {
      await this.client.send(
        new PutItemCommand({
          TableName: TABLE_NAME,
          Item: {
            UserId: { S: userId },
            Timestamp: { S: timestamp },
          },
        }),
      );
    } catch (error) {
      this.lastLog(`[DynamoDB] createUser error: ${error}`);
      throw error;
    }
  }

  async updateTimestamp(userId: string): Promise<void> {
    if (!this.client) throw new Error('DynamoDBService not initialized');

    const timestamp = this.nowTimestamp();
    this.lastLog(`[DynamoDB] Updating ${userId} timestamp to ${timestamp}`);

    try {
      await this.client.send(
        new UpdateItemCommand({
          TableName: TABLE_NAME,
          Key: {
            UserId: { S: userId },
          },
          UpdateExpression: 'SET #ts = :ts',
          ExpressionAttributeNames: {
            '#ts': 'Timestamp',
          },
          ExpressionAttributeValues: {
            ':ts': { S: timestamp },
          },
        }),
      );
    } catch (error) {
      this.lastLog(`[DynamoDB] updateTimestamp error: ${error}`);
      throw error;
    }
  }

  private nowTimestamp(): string {
    // Local time, not UTC — matches the Go app's time.Now().Format("20060102150405")
    // and the Flutter app's DateTime.now(). All devices must use the same convention
    // so that lexicographic timestamp comparison in the orchestrator works correctly.
    const now = new Date();
    const year = now.getFullYear();
    const month = String(now.getMonth() + 1).padStart(2, '0');
    const day = String(now.getDate()).padStart(2, '0');
    const hour = String(now.getHours()).padStart(2, '0');
    const minute = String(now.getMinutes()).padStart(2, '0');
    const second = String(now.getSeconds()).padStart(2, '0');

    return `${year}${month}${day}${hour}${minute}${second}`;
  }
}

export const dynamoDBService = new DynamoDBService();
